### PR TITLE
MPI-IO Python Bindings

### PR DESCRIPTION
**Pull request by hirschsn**
This PR implements python bindings for the bare MPI-IO functions in ESPResSo.

Since we came up with this deficit during our meeting last week and I had this interface lying around for quite some time, I decided to open a PR.

Description of changes:
 - Add a python interface for MPI-IO functionality in the core.

If it is decided that we want to merge this upstream, I can write docs etc.

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

### CODE REVIEW

File: src/python/espressomd/io/mpiio.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        velocities : :obj:`bool`, optional
+                    Indicates if velocities should be dumped.
+        types : :obj:`bool`, optional
+                      Indicates if types should be dumped.
+        bonds : :obj:`bool`, optional
+                     Indicates if bonds should be dumped.
+        """
+
+        if prefix is None:
+            print "[MPI-IO] Need to supply output prefix via 'prefix' kwarg."
```
Comment (KaiSzuttor) [inline src/python/espressomd/io/mpiio.py]:
``` python
raise ValueError("...")
```

File: src/python/espressomd/io/mpiio.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            return
+        if not positions and not velocities and not types and not bonds:
+            print "[MPI-IO] No output fields chosen. Not performing any write."
```
Comment (KaiSzuttor) [inline src/python/espressomd/io/mpiio.py]: raise

File: src/python/espressomd/io/mpiio.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        This function reads data dumped by `write'. See the write documentation
+        for details.
+
+        .. note::
+            The files must be read on the same number of processes that wrote
+            the data. The data must be read on a machine with the same
+            architecture (otherwise, this might silently fail).
+        """
+        if prefix is None:
+            print "[MPI-IO] Need to supply output prefix via 'prefix' kwarg."
```
Comment (KaiSzuttor) [inline src/python/espressomd/io/mpiio.py]: raise

File: src/python/espressomd/io/mpiio.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            return
+        if not positions and not velocities and not types and not bonds:
+            print "[MPI-IO] No output fields chosen. Not performing any read."
```
Comment (KaiSzuttor) [inline src/python/espressomd/io/mpiio.py]: raise

File: src/core/mpiio.cpp [hunk 5]
```diff
@@ -449,4 +449,6 @@ void mpi_mpiio_common_read(const char *filename, unsigned fields) {
     clear_particle_node();
 
   on_particle_change();
+  // Out of box particles might be accepted by the cell system.
+  set_resort_particles(Cells::RESORT_GLOBAL);
```
Comment (KaiSzuttor) [inline src/core/mpiio.cpp:455]: Can you please move all corresponding files into the io directory.

File: src/python/espressomd/io/mpiio.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+         - bond: Bond information (if dumped): variable amount of data,
+         - boff: Bond offset information (if bonds are dumped): 1 int per particle.
+
+        .. note::
+            Do not read the files on a machine with a different architecture!
+
+        Parameters
+        ----------
+        prefix : :obj:`str`
+                   Common prefix for the filenames.
```
Comment (KaiSzuttor) [inline src/python/espressomd/io/mpiio.py]: ~~just to be picky... please align the explanation with the type description above~~ please autopep8 -i on that file

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]
Looks good to me besides the comments. In the long run you may want to check using h5md. But there you would have to implement a reader, different story.

**Comment by hirschsn** [comment]
Moved the files, put in appropriate python errors, removed the old mpi call stuff from `communication.[ch]pp` and added and fixed the documentation.

File: doc/sphinx/io.rst [hunk 7]
```diff
@@ -239,6 +239,45 @@ h5.write()
 After the last write call, you have to call the close() method to remove
 the backup file and to close the datasets etc.
 
+Writing MPI-IO binary files
+---------------------------
+
+This method outputs binary data in parallel and is, thus, also suitable for
+large-scale simulations. Generally, H5MD is the preferred method because the
+data is easier accessible. In contrast to H5MD, the MPI-IO functionality
+outputs data in a *machine dependant format* but has write and read
```
Comment (KaiSzuttor) [inline doc/sphinx/io.rst]: typo: dependant

File: doc/sphinx/io.rst [hunk 8]
```diff
@@ -239,6 +239,45 @@ h5.write()
...
+    - mydata.id
+    - mydata.pos
+    - mydata.pref
+    - mydata.type
+    - mydata.vel
+    - mydata.boff
+    - mydata.bond
+
+Depending on the chosen output, not all of these files might be created.
+To read these in again, simply call ``mpiio.read``. It has the same signature as
```
Comment (KaiSzuttor) [inline doc/sphinx/io.rst]: can you please directly reference the method by using ```:mod:`espressomd.io.mpiio.read` ```

**Comment by hirschsn** [comment]
Thanks for looking over the patches, again! Fixed both problems.

**Review by KaiSzuttor — APPROVED** [review]

**Comment by KaiSzuttor** [comment]
Could you please merge the current python branch?

**Comment by KaiSzuttor** [comment]
``` bash
ImportError: /builds/espressomd/espresso/build/src/script_interface/libEspressoScriptInterface.so.4: undefined symbol: _ZN5Mpiio22mpi_mpiio_common_writeEPKcj
```

**Comment by hirschsn** [comment]
Sorry, forgot to include the changes to CMakeLists.txt in one of the commits.

**Review by KaiSzuttor — APPROVED** [review]

**Pull request merged**